gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/PartitionedRegion.java

```diff
@@ -7172,9 +7176,11 @@ public boolean hasNext() {
             // KN: If DiskIterator has been initialized then we need to change
             // the current bucket id and the current bucket region accordingly
             // as otherwise there will be a mismatch
-            if (val != null && this.diskIteratorInitialized) {
+            if (val != null && !(val instanceof NonLocalRegionEntry) && this.diskIteratorInitialized) {
```

Comment:
Checked that Neeraj added this for #51837 in eecc9937. Lets check with him if the reason for this and if skipping it is fine.